### PR TITLE
Move FieldConfigSource from fieldOptions to PanelModel.fieldConfig

### PR DESCRIPTION
**Pull request by dprokop**
~Includes cherry-picked 5ce796caa2e82845e9ad75568b1b912762370a74 from https://github.com/grafana/grafana/pull/22754~

TODO:
- [x] decide on what to do with fieldOptions property, should it be standardised on PanelModel level?
- [x] Fix GaugeMigrations
- [x] add `onFieldConfigChange` to PanelProps API

### CODE REVIEW

File: public/app/features/dashboard/state/PanelModel.ts
Comment (dprokop) [inline public/app/features/dashboard/state/PanelModel.ts]: Newly added model property for storing field config options

File: public/app/features/dashboard/state/PanelModel.ts
Comment (dprokop) [inline public/app/features/dashboard/state/PanelModel.ts]: Used by panel editor to modify field config

File: packages/grafana-data/src/types/panel.ts
Comment (dprokop) [inline packages/grafana-data/src/types/panel.ts]: Stores panel's field config, setter added below. Used i.e. in stat and piechart panel. Need to make sure the defaults are respected.

File: public/app/features/dashboard/state/DashboardMigrator.ts
Comment (dprokop) [inline public/app/features/dashboard/state/DashboardMigrator.ts]: removing defaults and overrides from "legacy" field options.

File: packages/grafana-data/src/field/fieldDisplay.ts
Comment (dprokop) [inline packages/grafana-data/src/field/fieldDisplay.ts]: Added field config here for title interpolation

File: public/app/plugins/panel/stat/StatPanel.tsx
Comment (dprokop) [inline public/app/plugins/panel/stat/StatPanel.tsx]: Not super wild about this, I'm considering moving the rest of fieldOptions into PanelModel as well, but don't have any strong argument for or against. Feels like this si standardised property, so probably could live in the PanelModel

Comment (dprokop) [reply public/app/plugins/panel/stat/StatPanel.tsx]: Currently standard field display options are rendered "together" with other field options that are custom for a given panel. Probably the standard ones (FieldDisplayEditor) could be rendered in a collapsible section as other standard config options we have, and only the custom options would be rendered below. Not sure - any thoughts around this @torkelo @ryantxu ?

**Comment by ryantxu** [comment]
This adds [fieldConfig](https://github.com/grafana/grafana/blob/fieldoptions-fieldconfig/packages/grafana-data/src/types/panel.ts#L39) to PanelProps, but not a way for panels to change the config directly.

I think the standard UI should handle most of it, but I also think plugins will need to be able to manipulate the field configs.

Should we also add:
```
  onFieldConfigChange: (config) => void;
```

File: public/app/plugins/panel/stat/StatPanel.tsx
Comment (ryantxu) [reply public/app/plugins/panel/stat/StatPanel.tsx]: I don't think the other parts should be on the PanelModel.  I think they can continue to exist as a "common" property used by many panels.  The complexity and logic is kinda limited.

I'm not sure the best name -- the remaining values are about how to select the fields for display

**Comment by dprokop** [comment]
@torkelo @ryantxu this is now ready for your review 🙏

File: packages/grafana-data/src/types/panel.ts
Comment (torkelo) [inline packages/grafana-data/src/types/panel.ts]: this feels a bit off, should it not be fieldConfigDefaults: FieldConfig 

Or maybe we want panels with default overrides I guess.  

thoguhts @ryantxu ?

File: public/app/features/dashboard/state/PanelModel.ts
Comment (torkelo) [inline public/app/features/dashboard/state/PanelModel.ts]: merging the overrides array, how does that work?

File: packages/grafana-data/src/types/panel.ts
Comment (torkelo) [inline packages/grafana-data/src/types/panel.ts]: a bit of a mismatch here property name & function name (in terms of type/param). Not sure what to with overrides, separate function & property? or change argument here to be FieldConfigSource

**Review by torkelo — COMMENTED** [review]
Looks good, some details that we need to discuss and brainstorm around defaults and overrides. 

But the biggest problem is we cannot merge this until after we switch to the new panel editor as this removed all the options from the existing panel editor. Not sure what interim solution we can provide to make this mergable before that. Restore the panel options but use onFieldConfigChange ? 

The migrations for bar gauge looks a bit broken (or something else is causing it) , same for stat panel 
image

File: packages/grafana-data/src/types/panel.ts
Comment (dprokop) [reply packages/grafana-data/src/types/panel.ts]: Good point, not really sure so I'll wait for Ryan to kick in.

**Comment by dprokop** [comment]
>The migrations for bar gauge looks a bit broken (or something else is causing it) , same for stat panel
@torkelo  this is fixed now

File: packages/grafana-data/src/types/panel.ts
Comment (dprokop) [reply packages/grafana-data/src/types/panel.ts]: @torkelo updated the API to accept FieldConfigSource. @ryantxu do you have eny thoughts on this if overrides should be also allowed to be set as panel's defaults?

File: public/app/features/dashboard/state/PanelModel.ts
Comment (dprokop) [reply public/app/features/dashboard/state/PanelModel.ts]: uuuups, it probably doesn't ;)

Comment (dprokop) [reply public/app/features/dashboard/state/PanelModel.ts]: Updated the logic and added test for the behaviour

File: packages/grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts
Comment (torkelo) [inline packages/grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts]: Since you added a migration to DashboardMigration for this (which I think was a good idea), do we need this migration here as well?

**Review by torkelo — COMMENTED** [review]
Great job, some tricky stuff here :) 

Think right now I think we are some of the old defaults, stat & bar gauge & gauge all used this I think. 
https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/stat/types.ts#L26

File: packages/grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts
Comment (dprokop) [reply packages/grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts]: @torkelo that's primarily because of the former migrations that happen in this migration handler

Comment (torkelo) [reply packages/grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts]: ok, but then it feels like we will end up in strange place, where DashboardMigrator runs first, so these old panel migrations will not work.

**Review by torkelo — APPROVED** [review]

**Comment by torkelo** [comment]
@ryantxu we are merging this now, but please glance through it if you have any additional thoughts. There are somethings that are temporary to support the old edit mode

**Pull request merged**
